src/factories/model.js

```diff
@@ -0,0 +1,267 @@
+     */
+    hasCollection(type) {
+        return this.collections.hasOwnProperty(type);
+    }
+
+    /**
+     * Instantiate a new Collection by type.
+     * @param {String} type The collection type.
+     * @param {Object} arr The initial array of models.
+     * @param {Function} BaseCollection The base collection constructor.
+     * @return {Promise<Collection>}
+     */
+    initCollection(type, arr, BaseCollection) {
+        let resolveCollectionCtr;
+        if (typeof data === 'function') {
+            // initModel({type}, {BaseCollection})
+            BaseCollection = arr;
+            arr = undefined;
+        }
+        if (typeof type === 'string') {
+            // get or generate the collection
+            resolveCollectionCtr = this.getCollection(type, BaseCollection);
+        } else if (typeof type === 'function') {
+            // already a Class contrusctor
+            resolveCollectionCtr = Promise.resolve(type);
+        }
+        return resolveCollectionCtr.then((Collection) => this.initClass(Collection, arr));
+    }
+
+    /**
+     * Instantiate a new Model by type.
+     * @param {String} type The model type.
+     * @param {Object} data A set of initial data.
+     * @param {Function} BaseModel The base model constructor.
+     * @return {Promise<Model>}
+     */
+    initModel(type, data, BaseModel) {
+        let resolveModelCtr;
+        if (typeof data === 'function') {
+            // initModel({type}, {BaseModel})
+            BaseModel = data;
+            data = undefined;
+        }
+        if (typeof type === 'string') {
+            // get or generate the model
+            resolveModelCtr = this.getModel(type, BaseModel);
+        } else if (typeof type === 'function') {
+            // already a Class contrusctor
+            resolveModelCtr = Promise.resolve(type);
+        }
+        return resolveModelCtr.then((Model) => this.initClass(Model, data));
+    }
+
+    /**
+     * Fecth all object types.
+     * @return {Promise<ObjectTypesCollection>}
+     */
+    getObjectTypes() {
+        return this.initClass(ObjectTypesCollection)
+            .then((collection) =>
+                collection.findAll()
+                    .then(() => collection)
+            );
+    }
+
+    /**
+     * Fecth abstract object types.
+     * @return {Promise<ObjectTypesCollection>}
+     */
+    getAbastractObjectTypes() {
+        return this.initClass(ObjectTypesCollection)
+            .then((collection) =>
+                collection.findAll({
+                    filter: {
+                        is_abstract: true,
+                    },
+                }).then(() => collection)
+            );
+    }
+
+    /**
+     * Get the ObjectType Model for a given type.
+     * @param {String} type The type of the Model.
+     * @return {Promise<ObjectType>}
+     */
+    getObjectType(type) {
+        return this.initClass(ObjectTypesCollection)
+            .then((collection) => collection.findById(type));
```

Comment:
`findById(type)` is right? I'd expect and `id` as parameter to `findById`

Reply:
In this case, the type name is a string identifier.